### ISSUE TITLE
[Feature]: Unauthenticated users should not be able to call the API

### ISSUE THREAD

**Issue by gcapizzi** [cloudfoundry/cf-k8s-controllers#232]
### Background

While we work to make sure that all our endpoints move from using privileged clients to user-authenticated clients, we want to introduce a basic layer of security to the API. Let's start by checking if the user calling is authenticated at all, even if the requests we'll make to Kubernetes will still be performed as a privileged user instead of the calling user.

### Acceptance Criteria

**GIVEN** I **am** logged in
**WHEN I** hit any* endpoint of the API shim
**THEN I** see the request has been **accepted**

**GIVEN** I **am not** logged in
**WHEN I** hit any* endpoint of the API shim
**THEN I** see the request has been **denied** with a response like this:
```
HTTP/1.1 401 Unauthorized
Content-Type: application/json
```
```json
{
  "errors": [
    {
      "code": 10002,
      "title": "CF-NotAuthenticated",
      "detail": "Authentication error"
    }
  ]
}
```

**GIVEN** I **am** logged in
**BUT** my token or cert is not valid on the cluster
**WHEN I** hit any* endpoint of the API shim
**THEN I** see the request has been **denied** with a response like this:
```
HTTP/1.1 403 Forbidden
Content-Type: application/json
```
```json
{
  "errors": [
    {
      "code": 10003,
      "title": "CF-NotAuthorized",
      "detail": "You are not authorized to perform the requested action"
    }
  ]
}
```

#### *Whitelisted endpoints

We want the following endpoints to always be accessible:
* `/`
* `/v3`

### Dev Notes

* Given we want to protect the whole API using the same logic, we could use a [`gorilla/mux` middleware](https://github.com/gorilla/mux#middleware).
* To verify if a user is authenticated or not, the simplest thing to do is probably to try performing a request to Kubernetes on their behalf. Ideally, the request shouldn't necessarily be one the user is authorised for, all we want is to distinguish users who are _authenticated_ from users who aren't (regardless of wether they're _authorized_, for now).
  - [k8s.io/apimachinery/pkg/api/errors.IsUnauthorized](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsUnauthorized) and [k8s.io/apimachinery/pkg/api/errors.IsForbidden](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsForbidden) could help here
* This is temporary, until we add authentication to all our endpoints.
* The distinction between _unauthenticated_ (`401`) and _unauthorised_ (`403`) is one that we probably want to maintain.

**Comment by gcapizzi** [comment]
As discussed today at IPM, we _should_ be able to use the `IndentityProvider` for this once we fix #240.

**Comment by gcapizzi** [comment]
We have found two ways to tell authenticated users (and signed certificates) apart:
1. Perform any action (e.g. listing namespaces, or even listing `Pod`s in the `default` namespace) and check the response.
    - If `401` (`IsUnauthorized`), the user is not authenticated, so the cert is invalid.
    - If `403` (`IsForbidden`), the user is authenticated but not permitted to perform the action.
1. Try to create a `SelfSubjectAccessReview` (or a `SelfSubjectRulesReview`), which are the resources created by the `kubectl auth can-i` command (see [_ Authorization Overview_](https://kubernetes.io/docs/reference/access-authn-authz/authorization/)).
    - If `201`, the user is authenticated and the cert is valid. Every user can create these resources.
    - If not, the user is not authenticated and so the cert is invalid.

The latter option feels a lot more elegant and safe.
